enlace/participants/all_impact.php

```diff
@@ -20,83 +20,123 @@
- * programs ('a') or if they have access to a program that is linked
- * to this...person? or this survey?: */
-
-$access_array = $USER->program_access($Enlace_id);
 
+$anonymous_entry = false;
 include "../include/dbconnopen.php";
 $person_sqlsafe=mysqli_real_escape_string($cnnEnlace, $_GET['person']);
-$get_program_list = "SELECT Session_Names.Program_ID FROM Participants_Programs INNER JOIN Session_Names ON Participants_Programs.Program_ID = Session_ID WHERE Participant_ID = '" . $person_sqlsafe . "'";
-$program_connected = mysqli_query($cnnEnlace, $get_program_list);
-$program_array = array();
-while ($program_id = mysqli_fetch_row($program_connected)){
-    $program_array[] = $program_id[0];
-}
-$USER->enforce_access_program_array($Enlace_id, $program_array);
-
+include "../include/dbconnopen.php";
 
-/*
- * This file is either a new impact survey, an editable impact survey, or a view for the impact survey.
- * The responses to questions are shown if they have been answered, and are always editable.
- */
+if (isset($_GET['l'])) {
+    include "../include/dbconnopen.php";
+    $session_id_safe = mysqli_real_escape_string($cnnEnlace, $_GET['session']);
+    include "../include/dbconnclose.php";
+    ensure_survey_entry($person_sqlsafe, $session_id_safe, $_GET['l'], 2);
+    $anonymous_entry = true;
+}
 
-//get participant info
-include "../classes/participant.php";
+if(!$anonymous_entry) {
+    include "../header.php";
+    user_enforce_has_access($Enlace_id);
+}
 
-include "../include/dbconnopen.php";
-$id_sqlsafe=mysqli_real_escape_string($cnnEnlace, $_GET['assessment']);
-$person_sqlsafe=mysqli_real_escape_string($cnnEnlace, $_GET['person']);
-/* if the assessment is not new, then Get['assessment'] exists, and this query returns the entered responses. */
-$get_assessment_info = "SELECT * FROM Assessments 
-                        LEFT JOIN Participants_Caring_Adults ON Caring_Id=Caring_Adults_ID
-                        LEFT JOIN Participants_Future_Expectations ON Future_Id=Future_Expectations_ID
-                        LEFT JOIN Participants_Interpersonal_Violence ON Violence_Id=Interpersonal_Violence_ID
-                        LEFT JOIN Programs ON Participants_Future_Expectations.Program=Programs.Program_ID
-                        WHERE Assessment_ID='" . $id_sqlsafe . "'";
-$get_assessment = mysqli_query($cnnEnlace, $get_assessment_info);
-$assessment_info = mysqli_fetch_array($get_assessment);
-$caring_id = $assessment_info['Caring_ID'];
-$baseline_id = $assessment_info['Baseline_ID'];
-$future_id = $assessment_info['Future_ID'];
-$violence_id = $assessment_info['Violence_ID'];
-$person = new Participant();
-$person->load_with_participant_id($assessment_info[1]);
+if(!$anonymous_entry) {
+    /* so allow them in if it's a new survey OR if they have access to all
+     * programs ('a') or if they have access to a program that is linked
+     * to this...person? or this survey?: */

```

Comment:
Does this question (which was copied over from a comment that was moved here from somewhere else above) get answered now with this change?

A general comment: It would really help to have, somewhere, a high-level description of the intended behavior.  I'm fuzzy on exactly what the usage scenario is for this non-logged-in survey access feature.  Someone who doesn't have a login account wants to fill out a survey?  Okay, then do they also want to be able to go back and view (or even edit) their filled-out survey later? If they do want the latter, would it work, given that the session might be different now?

In general, without a thorough and unambiguous description of the usage scenario(s), it's hard for me to review the code changes.

Reply:
The use case is that someone who does data entry is sitting down with a youth.  They want to open up a page (or email a link) to allow the youth to fill out the survey, rather than doing it in pdf or hard copy and then manually entering it.  Then they want to log out so the youth doesn't mess around on the site.

Reply:
As for documenting the feature, that's challenging.  It's kind of a wide ranging feature touching many of the files.  The best documentation is in the feature request (#221), and the commit message, I think.

I can totally see updating #221 if it's unclear, but right now there's no feature documentation place (which we've talked about) where this can be added.

Reply:
So wait: when did the browser get logged in in the first place, in that scenario?  That is, why not offer a unique link that lets someone fill out a survey but do nothing else, and no logging-in happens in the first place.